plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -807,8 +808,22 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
          my->chain_config->block_validation_mode = options.at("validation-mode").as<validation_mode>();
       }
 
+      // controller must have config at creation, and config must contain genesis hash
+      if (my->chain_config->read_genesis) {
+         const auto& file = my->chain_config->genesis_file;
+         EOS_ASSERT(fc::is_regular_file(file), plugin_config_exception,
+            "Genesis state file '${f}' does not exist.", ("f", file.generic_string()));
+         string content;
+         fc::read_file_contents(file, content);
```

Comment:
Read all file into memory may take too much time. Maybe map file into memory? Or check hash when read genesis data (in genesis_import)?

Reply:
Originally I wrote it in genesis_import, and it read file by 64k blocks. But we need hash much earlier than genesis_import created (actually, chain::controller instantiation already wants config with `genesis_state` struct, which contains hash).
Actually, `read_file_contents` used in fork_db initialization, so looks like it's fast enough, and I removed reading by blocks to make code simpler.
Current generic-genesis file is 110M, it reads instantly